Start of Capstone 2: loading order information for a company <br>
Orders will included different lines for each product in the order <br>
### Goal of this Notebook:
### 1. remove or mask all data that could be indendifying and cannot be seen by the public
### 2. combine mulitiple CSV files into 1 dataframe 
### 3. make sure we have a unique customer ID for each email address


In [39]:
import os
import pandas as pd
import numpy as np
import datetime

In [40]:
# change to the path with the raw csv file data
os.chdir("C:\Springboard\Github\Capstone2_cust\Raw_CSV") 

In [41]:
# load the first csv
df1 = pd.read_csv("orders_export_1.csv", low_memory=False)
# look at the first 10 rows of this file
df1.head(10)

Name                      Email Financial Status  \
0  #64088     sailing19046@gmail.com             paid   
1  #64087      erinemeehan@gmail.com             paid   
2  #64087      erinemeehan@gmail.com              NaN   
3  #64087      erinemeehan@gmail.com              NaN   
4  #64086  kimberlyray0120@gmail.com             paid   
5  #64086  kimberlyray0120@gmail.com              NaN   
6  #64085        khalasz79@gmail.com             paid   
7  #64085        khalasz79@gmail.com              NaN   
8  #64084   jodigauthier23@Gmail.com             paid   
9  #64084   jodigauthier23@Gmail.com              NaN   

                     Paid at Fulfillment Status Fulfilled at  \
0  2020-09-12 05:49:40 -0700        unfulfilled          NaN   
1  2020-09-12 05:36:26 -0700        unfulfilled          NaN   
2                        NaN                NaN          NaN   
3                        NaN                NaN          NaN   
4  2020-09-12 05:28:27 -0700        unfulfilled          NaN   
5                        NaN                NaN          NaN   
6  2020-09-12 05:26:56 -0700        unfulfilled          NaN   
7                        NaN                NaN          NaN   
8  2020-09-12 05:07:11 -0700        unfulfilled          NaN   
9                        NaN                NaN          NaN   

  Accepts Marketing Currency  Subtotal  Shipping  ...  Tax 2 Name  \
0                no      USD     25.00       0.0  ...         NaN   
1                no      USD     90.68       0.0  ...         NaN   
2               NaN      NaN       NaN       NaN  ...         NaN   
3               NaN      NaN       NaN       NaN  ...         NaN   
4               yes      USD     54.97       0.0  ...         NaN   
5               NaN      NaN       NaN       NaN  ...         NaN   
6                no      USD     54.97       5.0  ...         NaN   
7               NaN      NaN       NaN       NaN  ...         NaN   
8                no      USD     24.98       3.0  ...         NaN   
9               NaN      NaN       NaN       NaN  ...         NaN   

   Tax 2 Value Tax 3 Name  Tax 3 Value Tax 4 Name Tax 4 Value  Tax 5 Name  \
0          NaN        NaN          NaN        NaN         NaN         NaN   
1          NaN        NaN          NaN        NaN         NaN         NaN   
2          NaN        NaN          NaN        NaN         NaN         NaN   
3          NaN        NaN          NaN        NaN         NaN         NaN   
4          NaN        NaN          NaN        NaN         NaN         NaN   
5          NaN        NaN          NaN        NaN         NaN         NaN   
6          NaN        NaN          NaN        NaN         NaN         NaN   
7          NaN        NaN          NaN        NaN         NaN         NaN   
8          NaN        NaN          NaN        NaN         NaN         NaN   
9          NaN        NaN          NaN        NaN         NaN         NaN   

  Tax 5 Value  Phone  Receipt Number  
0         NaN    NaN             NaN  
1         NaN    NaN             NaN  
2         NaN    NaN             NaN  
3         NaN    NaN             NaN  
4         NaN    NaN             NaN  
5         NaN    NaN             NaN  
6         NaN    NaN             NaN  
7         NaN    NaN             NaN  
8         NaN    NaN             NaN  
9         NaN    NaN             NaN  

[10 rows x 72 columns]

Let's look at the columns in this DF

In [42]:
df1.columns

Index(['Name', 'Email', 'Financial Status', 'Paid at', 'Fulfillment Status',
       'Fulfilled at', 'Accepts Marketing', 'Currency', 'Subtotal', 'Shipping',
       'Taxes', 'Total', 'Discount Code', 'Discount Amount', 'Shipping Method',
       'Created at', 'Lineitem quantity', 'Lineitem name', 'Lineitem price',
       'Lineitem compare at price', 'Lineitem sku',
       'Lineitem requires shipping', 'Lineitem taxable',
       'Lineitem fulfillment status', 'Billing Name', 'Billing Street',
       'Billing Address1', 'Billing Address2', 'Billing Company',
       'Billing City', 'Billing Zip', 'Billing Province', 'Billing Country',
       'Billing Phone', 'Shipping Name', 'Shipping Street',
       'Shipping Address1', 'Shipping Address2', 'Shipping Company',
       'Shipping City', 'Shipping Zip', 'Shipping Province',
       'Shipping Country', 'Shipping Phone', 'Notes', 'Note Attributes',
       'Cancelled at', 'Payment Method', 'Payment Reference',
       'Refunded Amount', 'Vendor', '

In [43]:
# dimensions of the dateframe
df1.shape

(31498, 72)

In [44]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 31498 entries, 0 to 31497
Data columns (total 72 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   Name                         31498 non-null  object 
 1   Email                        31467 non-null  object 
 2   Financial Status             13900 non-null  object 
 3   Paid at                      13463 non-null  object 
 4   Fulfillment Status           13900 non-null  object 
 5   Fulfilled at                 13600 non-null  object 
 6   Accepts Marketing            13900 non-null  object 
 7   Currency                     13900 non-null  object 
 8   Subtotal                     13900 non-null  float64
 9   Shipping                     13900 non-null  float64
 10  Taxes                        13900 non-null  float64
 11  Total                        13900 non-null  float64
 12  Discount Code                1051 non-null   object 
 13  Discount Amount 

## Load the other CSV's and append them to that first Dataframe

In [45]:
# load the second csv
df2 = pd.read_csv("orders_export_2.csv", low_memory=False)
# look at the first 10 rows of this file
df2.head(10)

Name                        Email Financial Status  \
0  #50186  jeanine.alvarez@hotmail.com             paid   
1  #50185      wuest.heather@gmail.com             paid   
2  #50184        k8petteruto@gmail.com             paid   
3  #50183     sweetmoonriver@yahoo.com             paid   
4  #50183     sweetmoonriver@yahoo.com              NaN   
5  #50182           koneil27@gmail.com             paid   
6  #50181       Rambler0429@icloud.com             paid   
7  #50181       Rambler0429@icloud.com              NaN   
8  #50180        garner366@hotmail.com             paid   
9  #50179       katrinaaa147@gmail.com             paid   

                     Paid at Fulfillment Status               Fulfilled at  \
0  2020-07-01 00:21:16 -0700          fulfilled  2020-07-01 15:08:13 -0700   
1  2020-07-01 00:21:07 -0700          fulfilled  2020-07-01 15:10:36 -0700   
2  2020-07-01 00:21:07 -0700          fulfilled  2020-07-01 15:09:39 -0700   
3  2020-06-30 23:26:45 -0700          fulfilled  2020-07-01 15:07:44 -0700   
4                        NaN                NaN                        NaN   
5  2020-06-30 23:24:10 -0700          fulfilled  2020-07-01 15:08:13 -0700   
6  2020-06-30 23:16:55 -0700          fulfilled  2020-07-01 14:07:15 -0700   
7                        NaN                NaN                        NaN   
8  2020-06-30 23:16:05 -0700          fulfilled  2020-07-01 14:07:02 -0700   
9  2020-06-30 23:08:00 -0700          fulfilled  2020-07-01 14:05:37 -0700   

  Accepts Marketing Currency  Subtotal  Shipping  ...  Tax 2 Name  \
0                no      USD     64.80      0.00  ...         NaN   
1                no      USD    108.00      0.00  ...         NaN   
2               yes      USD     21.60      0.00  ...         NaN   
3                no      USD     65.97      0.00  ...         NaN   
4               NaN      NaN       NaN       NaN  ...         NaN   
5                no      USD     66.99      8.76  ...         NaN   
6                no      USD     54.97      0.00  ...         NaN   
7               NaN      NaN       NaN       NaN  ...         NaN   
8                no      USD     32.00      0.00  ...         NaN   
9               yes      USD     24.98      0.00  ...         NaN   

   Tax 2 Value Tax 3 Name  Tax 3 Value Tax 4 Name Tax 4 Value  Tax 5 Name  \
0          NaN        NaN          NaN        NaN         NaN         NaN   
1          NaN        NaN          NaN        NaN         NaN         NaN   
2          NaN        NaN          NaN        NaN         NaN         NaN   
3          NaN        NaN          NaN        NaN         NaN         NaN   
4          NaN        NaN          NaN        NaN         NaN         NaN   
5          NaN        NaN          NaN        NaN         NaN         NaN   
6          NaN        NaN          NaN        NaN         NaN         NaN   
7          NaN        NaN          NaN        NaN         NaN         NaN   
8          NaN        NaN          NaN        NaN         NaN         NaN   
9          NaN        NaN          NaN        NaN         NaN         NaN   

  Tax 5 Value         Phone  Receipt Number  
0         NaN           NaN             NaN  
1         NaN           NaN             NaN  
2         NaN           NaN             NaN  
3         NaN           NaN             NaN  
4         NaN           NaN             NaN  
5         NaN           NaN             NaN  
6         NaN           NaN             NaN  
7         NaN           NaN             NaN  
8         NaN           NaN             NaN  
9         NaN  1.262865e+10             NaN  

[10 rows x 72 columns]

In [46]:
df1 = df1.append(df2)
df1.shape

(67474, 72)

In [47]:
# load the third csv
df3 = pd.read_csv("orders_export_3.csv", low_memory=False)
# look at the first 10 rows of this file
df3.head(10)

Name                       Email Financial Status  \
0  #34636        gstoukides@gmail.com             paid   
1  #34636        gstoukides@gmail.com              NaN   
2  #34635  allisonkaip126@hotmail.com             paid   
3  #34635  allisonkaip126@hotmail.com              NaN   
4  #34634       serenalove8@gmail.com             paid   
5  #34633   nicole.haddad84@gmail.com             paid   
6  #34633   nicole.haddad84@gmail.com              NaN   
7  #34633   nicole.haddad84@gmail.com              NaN   
8  #34632  loganbrady110505@gmail.com             paid   
9  #34631      sjkempnelson@gmail.com             paid   

                     Paid at Fulfillment Status               Fulfilled at  \
0  2020-04-10 15:37:08 -0700          fulfilled  2020-04-13 11:38:49 -0700   
1                        NaN                NaN                        NaN   
2  2020-04-10 15:30:45 -0700          fulfilled  2020-04-13 11:38:36 -0700   
3                        NaN                NaN                        NaN   
4  2020-04-10 15:11:26 -0700          fulfilled  2020-04-13 11:38:36 -0700   
5  2020-04-10 15:09:08 -0700          fulfilled  2020-04-13 11:38:49 -0700   
6                        NaN                NaN                        NaN   
7                        NaN                NaN                        NaN   
8  2020-04-10 15:08:59 -0700          fulfilled  2020-04-13 11:38:37 -0700   
9  2020-04-10 15:03:41 -0700          fulfilled  2020-04-13 11:38:53 -0700   

  Accepts Marketing Currency  Subtotal  Shipping  ...  Tax 2 Name  \
0               yes      USD     24.98       0.0  ...         NaN   
1               NaN      NaN       NaN       NaN  ...         NaN   
2                no      USD     56.00       0.0  ...         NaN   
3               NaN      NaN       NaN       NaN  ...         NaN   
4                no      USD     58.50       0.0  ...         NaN   
5               yes      USD     70.58       0.0  ...         NaN   
6               NaN      NaN       NaN       NaN  ...         NaN   
7               NaN      NaN       NaN       NaN  ...         NaN   
8               yes      USD     53.99       0.0  ...         NaN   
9                no      USD     46.98       0.0  ...         NaN   

   Tax 2 Value Tax 3 Name  Tax 3 Value Tax 4 Name Tax 4 Value  Tax 5 Name  \
0          NaN        NaN          NaN        NaN         NaN         NaN   
1          NaN        NaN          NaN        NaN         NaN         NaN   
2          NaN        NaN          NaN        NaN         NaN         NaN   
3          NaN        NaN          NaN        NaN         NaN         NaN   
4          NaN        NaN          NaN        NaN         NaN         NaN   
5          NaN        NaN          NaN        NaN         NaN         NaN   
6          NaN        NaN          NaN        NaN         NaN         NaN   
7          NaN        NaN          NaN        NaN         NaN         NaN   
8          NaN        NaN          NaN        NaN         NaN         NaN   
9          NaN        NaN          NaN        NaN         NaN         NaN   

  Tax 5 Value         Phone  Receipt Number  
0         NaN  1.617839e+10             NaN  
1         NaN  1.617839e+10             NaN  
2         NaN           NaN             NaN  
3         NaN           NaN             NaN  
4         NaN           NaN             NaN  
5         NaN           NaN             NaN  
6         NaN           NaN             NaN  
7         NaN           NaN             NaN  
8         NaN           NaN             NaN  
9         NaN           NaN             NaN  

[10 rows x 72 columns]

In [48]:
# add this third dataframe to the first one
df1 = df1.append(df3)
df1.shape

(103218, 72)

In [49]:
# load the third csv
df4 = pd.read_csv("orders_export_4.csv", low_memory=False)
# look at the first 10 rows of this file
df4.head(10)

Name                                Email Financial Status  \
0  #18736                  jpauszek1@yahoo.com             paid   
1  #18736                  jpauszek1@yahoo.com              NaN   
2  #18736                  jpauszek1@yahoo.com              NaN   
3  #18735  rosangelis.colongonzalez1@gmail.com         refunded   
4  #18735  rosangelis.colongonzalez1@gmail.com              NaN   
5  #18735  rosangelis.colongonzalez1@gmail.com              NaN   
6  #18734                shelbylynn411@msn.com             paid   
7  #18734                shelbylynn411@msn.com              NaN   
8  #18734                shelbylynn411@msn.com              NaN   
9  #18733                anna.hartis@yahoo.com             paid   

                     Paid at Fulfillment Status               Fulfilled at  \
0  2019-12-21 14:24:27 -0800          fulfilled  2019-12-28 11:41:13 -0800   
1                        NaN                NaN                        NaN   
2                        NaN                NaN                        NaN   
3  2019-12-21 14:21:07 -0800          fulfilled  2019-12-28 11:38:35 -0800   
4                        NaN                NaN                        NaN   
5                        NaN                NaN                        NaN   
6  2019-12-21 14:10:48 -0800          fulfilled  2019-12-28 11:37:04 -0800   
7                        NaN                NaN                        NaN   
8                        NaN                NaN                        NaN   
9  2019-12-21 13:52:35 -0800          fulfilled  2019-12-28 11:39:44 -0800   

  Accepts Marketing Currency  Subtotal  Shipping  ...  Tax 2 Name  \
0                no      USD     50.98       0.0  ...         NaN   
1               NaN      NaN       NaN       NaN  ...         NaN   
2               NaN      NaN       NaN       NaN  ...         NaN   
3               yes      USD     27.00       0.0  ...         NaN   
4               NaN      NaN       NaN       NaN  ...         NaN   
5               NaN      NaN       NaN       NaN  ...         NaN   
6                no      USD     48.98       0.0  ...         NaN   
7               NaN      NaN       NaN       NaN  ...         NaN   
8               NaN      NaN       NaN       NaN  ...         NaN   
9               yes      USD     56.00       0.0  ...         NaN   

   Tax 2 Value Tax 3 Name  Tax 3 Value Tax 4 Name Tax 4 Value  Tax 5 Name  \
0          NaN        NaN          NaN        NaN         NaN         NaN   
1          NaN        NaN          NaN        NaN         NaN         NaN   
2          NaN        NaN          NaN        NaN         NaN         NaN   
3          NaN        NaN          NaN        NaN         NaN         NaN   
4          NaN        NaN          NaN        NaN         NaN         NaN   
5          NaN        NaN          NaN        NaN         NaN         NaN   
6          NaN        NaN          NaN        NaN         NaN         NaN   
7          NaN        NaN          NaN        NaN         NaN         NaN   
8          NaN        NaN          NaN        NaN         NaN         NaN   
9          NaN        NaN          NaN        NaN         NaN         NaN   

  Tax 5 Value         Phone  Receipt Number  
0         NaN           NaN             NaN  
1         NaN           NaN             NaN  
2         NaN           NaN             NaN  
3         NaN  1.939320e+10             NaN  
4         NaN  1.939320e+10             NaN  
5         NaN  1.939320e+10             NaN  
6         NaN           NaN             NaN  
7         NaN           NaN             NaN  
8         NaN           NaN             NaN  
9         NaN  1.864478e+10             NaN  

[10 rows x 72 columns]

In [50]:
# add this third dataframe to the first one
df1 = df1.append(df4)
df1.shape

(139130, 72)

In [51]:
# load the third csv
df5 = pd.read_csv("orders_export_5.csv", low_memory=False)
# look at the first 10 rows of this file
df5.head(10)

Name                        Email Financial Status  \
0  #2436         teeagustin@yahoo.com             paid   
1  #2436         teeagustin@yahoo.com              NaN   
2  #2436         teeagustin@yahoo.com              NaN   
3  #2436         teeagustin@yahoo.com              NaN   
4  #2435           lpgrow@hotmail.com             paid   
5  #2435           lpgrow@hotmail.com              NaN   
6  #2434            nbartlette@me.com             paid   
7  #2433  marcantonioandrea@gmail.com             paid   
8  #2433  marcantonioandrea@gmail.com              NaN   
9  #2433  marcantonioandrea@gmail.com              NaN   

                     Paid at Fulfillment Status               Fulfilled at  \
0  2016-11-26 12:44:48 -0800          fulfilled  2016-12-01 16:44:34 -0800   
1                        NaN                NaN                        NaN   
2                        NaN                NaN                        NaN   
3                        NaN                NaN                        NaN   
4  2016-11-26 12:19:23 -0800          fulfilled  2016-12-02 11:29:08 -0800   
5                        NaN                NaN                        NaN   
6  2016-11-26 11:50:29 -0800          fulfilled  2016-11-30 10:42:23 -0800   
7  2016-11-26 09:33:26 -0800          fulfilled  2016-12-02 12:13:48 -0800   
8                        NaN                NaN                        NaN   
9                        NaN                NaN                        NaN   

  Accepts Marketing Currency  Subtotal  Shipping  ...              Tax 2 Name  \
0                no      USD      81.6       0.0  ...  Orange County Tax 0.5%   
1               NaN      NaN       NaN       NaN  ...                     NaN   
2               NaN      NaN       NaN       NaN  ...                     NaN   
3               NaN      NaN       NaN       NaN  ...                     NaN   
4               yes      USD      89.6       0.0  ...  Orange County Tax 0.5%   
5               NaN      NaN       NaN       NaN  ...                     NaN   
6               yes      USD      14.4       0.0  ...  Orange County Tax 0.5%   
7               yes      USD      92.0       0.0  ...  Orange County Tax 0.5%   
8               NaN      NaN       NaN       NaN  ...                     NaN   
9               NaN      NaN       NaN       NaN  ...                     NaN   

   Tax 2 Value Tax 3 Name  Tax 3 Value Tax 4 Name Tax 4 Value  Tax 5 Name  \
0         0.41        NaN          NaN        NaN         NaN         NaN   
1          NaN        NaN          NaN        NaN         NaN         NaN   
2          NaN        NaN          NaN        NaN         NaN         NaN   
3          NaN        NaN          NaN        NaN         NaN         NaN   
4         0.45        NaN          NaN        NaN         NaN         NaN   
5          NaN        NaN          NaN        NaN         NaN         NaN   
6         0.07        NaN          NaN        NaN         NaN         NaN   
7         0.46        NaN          NaN        NaN         NaN         NaN   
8          NaN        NaN          NaN        NaN         NaN         NaN   
9          NaN        NaN          NaN        NaN         NaN         NaN   

  Tax 5 Value  Phone  Receipt Number  
0         NaN    NaN             NaN  
1         NaN    NaN             NaN  
2         NaN    NaN             NaN  
3         NaN    NaN             NaN  
4         NaN    NaN             NaN  
5         NaN    NaN             NaN  
6         NaN    NaN             NaN  
7         NaN    NaN             NaN  
8         NaN    NaN             NaN  
9         NaN    NaN             NaN  

[10 rows x 72 columns]

In [52]:
# add this third dataframe to the first one
df1 = df1.append(df5)
df1.shape

(142224, 72)

#### Now, we have one big dataframe to work with

In [53]:
# let's make sure indices are unique
df1.index.unique()

Int64Index([    0,     1,     2,     3,     4,     5,     6,     7,     8,
                9,
            ...
            35966, 35967, 35968, 35969, 35970, 35971, 35972, 35973, 35974,
            35975],
           dtype='int64', length=35976)

There are 142k rows and 36k indices <br>
Those are not unique; they must have been repeated with every new dataframe that was loaded. <br>
Let's reset the index <br>

In [54]:
df1.reset_index(drop=True, inplace=True)

The index looks much better. <br>
<br>
Since we will be using the email address as a customer ID, let's drop them all to lowercase

In [55]:
# changing email to lower case
df1['Email'] = df1['Email'].str.lower()

Before we anonymize the email addresses, let's extract the email server.

In [56]:
## add email server as feature
df1['Server'] = df1['Email'].str.extract(r'\@(.*)')

In [57]:
df1['Server'].value_counts().head(25)

gmail.com         75791
yahoo.com         27213
hotmail.com       10285
aol.com            5538
icloud.com         2652
comcast.net        2029
outlook.com        1470
msn.com            1454
live.com           1234
me.com              992
sbcglobal.net       981
att.net             824
cox.net             694
ymail.com           649
verizon.net         485
mac.com             340
rocketmail.com      292
charter.net         269
bambuearth.com      264
bellsouth.net       215
optonline.net       173
4x400.com           171
earthlink.net       165
aim.com              99
windstream.net       92
Name: Server, dtype: int64

We just noticed that in the top 25 email address, both the vendor and the holding company email servers show up. We should convert those accordingly, so they are hidden <br>


In [58]:
df1['Server'].replace(to_replace='bambuearth.com', value='vendor', inplace=True)
df1['Server'].replace(to_replace='4x400.com', value='hold', inplace=True)

In [59]:
# number of unique email domains
df1['Server'].unique().shape

(1611,)

All of the email addresses that are not in the top 25 would have less than 100 orders or less than 0.1% of the total orders. Each of these servers would be outliers in building a model feature. They are not useful for modeling. <br>
More importantly, some of these domain names could be used to identify small companies that have their own domain name email addresses. We are going to replace all of these with "custom" <br>

In [60]:
# extract the most popular email domains and put them in a list
domain25 = df1['Server'].value_counts().head(25).index.tolist()
domain25

['gmail.com',
 'yahoo.com',
 'hotmail.com',
 'aol.com',
 'icloud.com',
 'comcast.net',
 'outlook.com',
 'msn.com',
 'live.com',
 'me.com',
 'sbcglobal.net',
 'att.net',
 'cox.net',
 'ymail.com',
 'verizon.net',
 'mac.com',
 'rocketmail.com',
 'charter.net',
 'vendor',
 'bellsouth.net',
 'optonline.net',
 'hold',
 'earthlink.net',
 'aim.com',
 'windstream.net']

In [61]:
# create a list of all unique domains
domains = df1['Server'].unique().tolist()
domains

['gmail.com',
 'aol.com',
 'icloud.com',
 'yahoo.com',
 'drjennifer.com',
 'charter.net',
 'outlook.com',
 'hotmail.com',
 'att.net',
 'rbncollective.com',
 'me.com',
 'gpaea.org',
 'live.com',
 'verizon.net',
 'msn.com',
 'rocketmail.com',
 'bubleschampagne.com',
 'cox.net',
 'bellsouth.net',
 'sbcglobal.net',
 'ymail.com',
 'comcast.net',
 'optonline.net',
 'vendor',
 'turcofamily.net',
 'mac.com',
 'philander.edu',
 'powersvitality.com',
 'wowway.com',
 'commonthreadco.com',
 'rightsideup.co',
 'hotmsil.com',
 nan,
 'lkfreeman.com',
 'united.net',
 'rsu13.org',
 'kw.com',
 'jagmail.southalabama.edu',
 'sncglobal.net',
 '12gagephotography.com',
 'drake.edu',
 'maine.edu',
 'earthlink.net',
 'stvincent.edu',
 'everitedoor.com',
 'vols.utk.edu',
 'theadvocate.com',
 'chartermi.net',
 'mchsi.com',
 'indy.rr.com',
 'perfectmelodyhomecare.com',
 'frontier.com',
 'camtel.net',
 'adp.com',
 'yahoo.de',
 'pgcps.org',
 'bethelu.edu',
 'kent.edu',
 'bethel.edu',
 'fiu.edu',
 'windstream.net',


In [62]:
# remove the top 25 from the list of unique domains using list comprehension
dom = [x for x in domains if x not in domain25]

In [63]:
# use the remaining list of unique domains to replace with value='custom'
df1['Server'].replace(to_replace=dom, value='custom', inplace=True)

Now, we need to remove some of the other columns that are identifiers

In [64]:
# before dropping address information, let's create a boolean feature if Shipping Zip = Billing Zip
df1['ship_bill'] = df1['Shipping Zip'] == df1['Billing Zip']

In [65]:
# let's look at shipping phone values
df1['Shipping Phone'].unique()

array([nan, '(803) 614-5081', '(336) 661-5334', ..., '8067296739',
       '6304871121', '315-345-8209'], dtype=object)

It looks like I can take the first 3 numbers to use as the area code using Regex if I wanted to extract the area code

In [66]:
# let's look at the number of values for Shipping phones
df1['Shipping Phone'].unique().shape

(29703,)

In [67]:
# let's extract the area code for the Shipping phone number to leave an area of indication
## add area code as a feature
df1['Area_Code'] = df1['Shipping Phone'].str.extract(r'^\(*([2-9]\d{2})')

In [68]:
df1['Area_Code'].value_counts()

949    733
714    490
760    225
214    221
720    220
      ... 
582      1
769      1
932      1
669      1
986      1
Name: Area_Code, Length: 373, dtype: int64

In [69]:
# let's see if this filled all of the values
df1['Area_Code'].isnull().sum()

116150

That's a lot of missing values. Let's use the billing phone number to fill some of those

In [70]:
# extract billing area code first
df1['Bill_Area_Code'] = df1['Billing Phone'].str.extract(r'^\(*([2-9]\d{2})')
df1['Area_Code'].fillna(value=df1['Bill_Area_Code'], inplace=True)
df1['Area_Code'].isnull().sum()

115818

That helped a little, but I think we can fill some more with the last "Phone" column

In [71]:
# extract area code from general phone number
df1['Phone'] = str(df1['Phone'])
df1['Bill_Area_Code'] = df1['Phone'].str.extract(r'^\(*([2-9]\d{2})')
df1['Area_Code'].fillna(value=df1['Bill_Area_Code'], inplace=True)
df1['Area_Code'].isnull().sum()

115818

Maybe that's as good as we can get

In [72]:
# drop the shipping and billing information, except for the shipping zip code
df1.drop(['Billing Name', 'Billing Street', 'Billing Address1', 'Billing Address2', 'Billing Company',
       'Billing City', 'Billing Zip', 'Billing Province', 'Billing Country', 'Billing Phone', 'Shipping Name', 'Shipping Street',
       'Shipping Address1', 'Shipping Address2', 'Shipping Company', 'Shipping City', 'Shipping Province',
       'Shipping Country', 'Shipping Phone', 'Bill_Area_Code'], axis=1, inplace=True)

In [73]:
# there may still be identifying features; potentially in vendor
df1['Vendor'].value_counts()

Bambu Earth    106103
Route           36101
Name: Vendor, dtype: int64

In [74]:
# these need to be renamed to keep it anonymous
df1['Vendor'].replace(to_replace='Bambu Earth', value=0, inplace=True)
df1['Vendor'].replace(to_replace='Route', value=1, inplace=True)

In [75]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 142224 entries, 0 to 142223
Data columns (total 56 columns):
 #   Column                       Non-Null Count   Dtype  
---  ------                       --------------   -----  
 0   Name                         142224 non-null  object 
 1   Email                        141639 non-null  object 
 2   Financial Status             63084 non-null   object 
 3   Paid at                      61027 non-null   object 
 4   Fulfillment Status           63084 non-null   object 
 5   Fulfilled at                 62270 non-null   object 
 6   Accepts Marketing            63084 non-null   object 
 7   Currency                     63084 non-null   object 
 8   Subtotal                     63084 non-null   float64
 9   Shipping                     63084 non-null   float64
 10  Taxes                        63084 non-null   float64
 11  Total                        63084 non-null   float64
 12  Discount Code                7212 non-null    object 
 13 

In [76]:
# look at employee column to make sure that is not giving away any information
df1['Employee'].value_counts()

Kohlman Verheyen    1081
Amber Hawthorne      197
Marya Nordenhok      150
Donna Vota            70
Ashley Boyd           13
Matthew Faris          2
Name: Employee, dtype: int64

In [77]:
# since they are actual employee names, we should convert them to booleans
df1['Employee'] = df1['Employee'].notnull()
# this should also replace all of the null values with False

In [78]:
# phone number also looks like it could be something that could identify people
df1['Phone'].value_counts()

0        NaN\n1        NaN\n2        NaN\n3        NaN\n4        NaN\n          ..\n142219   NaN\n142220   NaN\n142221   NaN\n142222   NaN\n142223   NaN\nName: Phone, Length: 142224, dtype: float64    142224
Name: Phone, dtype: int64

In [79]:
# those are definitely real phone numbers. We already extracted area code from shipping phone (which was a more extensive list).
# we can drop this column
df1.drop(['Phone'], axis=1, inplace=True)

#### let's create customer ID's for each email address and then anonymize the email addresses

In [80]:
df1['Email'].value_counts()

amber@bambuearth.com       167
vnscheller@yahoo.com       134
akcrawford24@yahoo.com     113
phy.melissa@gmail.com      109
zmindyb@icloud.com          99
                          ... 
karlihatch25@gmail.com       1
hkygram@yahoo.com            1
kchambers17@gmail.com        1
tgauthiermail@yahoo.com      1
lsanch210@outlook.com        1
Name: Email, Length: 39679, dtype: int64

We have 39,679 unique emails <br>
Now, let's get this started. We'll search through the emails and make sure each email address has the same Cust_ID, so we can better track customers.

Let's actually look at using the existing customer ID's

In [81]:
df1['Id'].unique().shape

(63085,)

In [82]:
df1['Id'].value_counts()

1.709389e+12    1
2.637124e+12    1
2.637107e+12    1
2.637103e+12    1
2.637101e+12    1
               ..
1.981126e+12    1
1.981120e+12    1
1.981117e+12    1
1.981116e+12    1
2.542631e+12    1
Name: Id, Length: 63084, dtype: int64

Those look like they could be pretty good as identifiers, but to not mess with that column, we'll use the index

In [83]:
# uses the first Id that appears by an email address as that unique identifier
df1['Cust_ID'] = df1.groupby('Email')['Id'].transform('first')
df1['Cust_ID'].value_counts()

2.747612e+12    167
2.599750e+12    134
2.746662e+12    113
2.602749e+12    109
2.647921e+12     99
               ... 
2.010881e+12      1
2.010797e+12      1
1.828734e+12      1
5.091840e+11      1
2.069237e+12      1
Name: Cust_ID, Length: 39679, dtype: int64

The unique customer ID is equal to the number of email addresses. That's exactly what we wanted.

In [84]:
# make sure that Cust_ID are the same when the email addresses are the same
df2 = df1
# we look for any cases where email addresses are the same but Cust_ID are different
df1[(df1['Email'] == df2['Email']) & (df1['Cust_ID'] != df2['Cust_ID'])]

Empty DataFrame
Columns: [Name, Email, Financial Status, Paid at, Fulfillment Status, Fulfilled at, Accepts Marketing, Currency, Subtotal, Shipping, Taxes, Total, Discount Code, Discount Amount, Shipping Method, Created at, Lineitem quantity, Lineitem name, Lineitem price, Lineitem compare at price, Lineitem sku, Lineitem requires shipping, Lineitem taxable, Lineitem fulfillment status, Shipping Zip, Notes, Note Attributes, Cancelled at, Payment Method, Payment Reference, Refunded Amount, Vendor, Outstanding Balance, Employee, Location, Device ID, Id, Tags, Risk Level, Source, Lineitem discount, Tax 1 Name, Tax 1 Value, Tax 2 Name, Tax 2 Value, Tax 3 Name, Tax 3 Value, Tax 4 Name, Tax 4 Value, Tax 5 Name, Tax 5 Value, Receipt Number, Server, ship_bill, Area_Code, Cust_ID]
Index: []

[0 rows x 56 columns]

That's extactly what I want to see - no duplicate customer numbers for the same email address <br>
Just to see if it was necessary, let's check their existing customer ID

In [85]:
df1['Cust_ID'].value_counts()

2.747612e+12    167
2.599750e+12    134
2.746662e+12    113
2.602749e+12    109
2.647921e+12     99
               ... 
2.010881e+12      1
2.010797e+12      1
1.828734e+12      1
5.091840e+11      1
2.069237e+12      1
Name: Cust_ID, Length: 39679, dtype: int64

In [86]:
df1['Cust_ID'].dtype

dtype('float64')

In [87]:
df1['Cust_ID'].isnull().sum()

585

In [88]:
# let's fill any missing values
df1['Cust_ID'].fillna(value=df1.index.to_series(), inplace=True)

In [89]:
df1['Cust_ID'].value_counts()

2.747612e+12    167
2.599750e+12    134
2.746662e+12    113
2.602749e+12    109
2.647921e+12     99
               ... 
2.013875e+12      1
2.013720e+12      1
2.013702e+12      1
2.013673e+12      1
2.069237e+12      1
Name: Cust_ID, Length: 40264, dtype: int64

In [90]:
#let's make sure they are all integers
df1['Cust_ID'].astype('int').value_counts()

-2147483648    140416
 2118104583        40
 2023075777        34
 1233196673        23
 2118262023        18
                ...  
 544496833          1
 134283             1
 134027             1
 131978             1
 131264             1
Name: Cust_ID, Length: 1016, dtype: int64

This looks really weird. I'm not going to convert to int

In [91]:
# let's make sure all of the new Cust_ID are filled
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 142224 entries, 0 to 142223
Data columns (total 56 columns):
 #   Column                       Non-Null Count   Dtype  
---  ------                       --------------   -----  
 0   Name                         142224 non-null  object 
 1   Email                        141639 non-null  object 
 2   Financial Status             63084 non-null   object 
 3   Paid at                      61027 non-null   object 
 4   Fulfillment Status           63084 non-null   object 
 5   Fulfilled at                 62270 non-null   object 
 6   Accepts Marketing            63084 non-null   object 
 7   Currency                     63084 non-null   object 
 8   Subtotal                     63084 non-null   float64
 9   Shipping                     63084 non-null   float64
 10  Taxes                        63084 non-null   float64
 11  Total                        63084 non-null   float64
 12  Discount Code                7212 non-null    object 
 13 

In [92]:
df1[df1['Cust_ID'].isnull()]

Empty DataFrame
Columns: [Name, Email, Financial Status, Paid at, Fulfillment Status, Fulfilled at, Accepts Marketing, Currency, Subtotal, Shipping, Taxes, Total, Discount Code, Discount Amount, Shipping Method, Created at, Lineitem quantity, Lineitem name, Lineitem price, Lineitem compare at price, Lineitem sku, Lineitem requires shipping, Lineitem taxable, Lineitem fulfillment status, Shipping Zip, Notes, Note Attributes, Cancelled at, Payment Method, Payment Reference, Refunded Amount, Vendor, Outstanding Balance, Employee, Location, Device ID, Id, Tags, Risk Level, Source, Lineitem discount, Tax 1 Name, Tax 1 Value, Tax 2 Name, Tax 2 Value, Tax 3 Name, Tax 3 Value, Tax 4 Name, Tax 4 Value, Tax 5 Name, Tax 5 Value, Receipt Number, Server, ship_bill, Area_Code, Cust_ID]
Index: []

[0 rows x 56 columns]

In [93]:
df1['Cust_ID'].dtype

dtype('float64')

In [94]:
# let's check the value counts for the customer ID
df1['Cust_ID'].value_counts()

2.747612e+12    167
2.599750e+12    134
2.746662e+12    113
2.602749e+12    109
2.647921e+12     99
               ... 
2.013875e+12      1
2.013720e+12      1
2.013702e+12      1
2.013673e+12      1
2.069237e+12      1
Name: Cust_ID, Length: 40264, dtype: int64

In [95]:
# extract the email and customer id to a separate DF that we can reference if necessary
lookup = df1[['Cust_ID', 'Email']].copy()

In [96]:
# let's remove the email column from the main dataframe so that the public doesn't see it
df1.drop(labels='Email', inplace=True, axis=1)

In [97]:
#let's export this data to work on in the next notebook
import pickle
os.chdir("C:\Springboard\Github\Capstone2_cust\Intermediate_Data")

In [98]:
# let's save the DF1 to both pickle and CSV
df1.to_pickle("cust_pub4.pkl")
df1.to_csv("cust_pub4.csv")

In [99]:
# let's also save the lookup DF
lookup.to_pickle("lookup4_priv.pkl")
lookup.to_csv("lookup4_priv.csv")

The End. More data pre-processing will take place in the next notebook, but this one needs to be private as we had to deal with anonymizing some of the data before making it public.